### PR TITLE
Microwave improvements

### LINKED ISSUE #7244: Microwaves should be able to hold and cook multiple items at once.
### Description
Microwaves should be able to hold and cook more than one item at once in order to make cooking less tedious. In /tg/ microwaves can hold 10 items at once by default. This amount can be modified depending on the tier of the matter bin used in the microwave.

N = 10 * t
In the above equation:
- N = Number of items the microwave can hold at once.
- t = Tier of the matter bin in the microwave.

### LINKED ISSUE #7281: Microwaves take twice the time they say they do when cooking food.
### Environment
- Game Version: Staging 21081519
- OS Version: Windows 10

### Detailed description of bug
Title. When a microwave is examined right after you start cooking with it, it will say it will take 10 seconds for it to cook. In actuality however, it takes 20 seconds to cook. Therefore, each microwave second is actually 2 seconds. You can see this in action by repeatedly examining a microwave as it cooks. It takes noticeably long for the microwave to go from saying 8 seconds are left to 7 seconds are left, and so on. 

### Steps to reproduce

1. Put food into a microwave, Turn on the microwave to start cooking.
2. Examine the microwave right after it's turned on, it will say cooking will take 10 seconds.
3. Actually wait 20 seconds for the food to finish cooking.

### PR DESCRIPTION
**Pull request by NoooneyDude**
- CL: [Feature] Microwaves will now work faster if overvolted.
- CL: [Balance] Microwaves have a chance to break by overvolting or cooking non-cookables. Just needs a screwdriver to fix.
- CL: [Balance] Microwaves now use power depending on current state and stock parts.
- CL: [Fix] Fixes microwave sprite init and recognizing stock parts. Fixes #7244 and fixes #7281.
- CL: [Feature] Multitools will now report the voltage and wattage of targeted APC-powered machines.
- CL: [Fix] Fixes bear meat and burned messes disappearing when cooked. Might be other foods that are problematic.
Added a warning to `Cookable` component to avoid setting the product item as itself.
- More flexible method to determine which methods an item can be cooked by for the `Cookable` component.

### CODE REVIEW

File: UnityProject/Assets/Scripts/Objects/Engineering/APCPoweredDevice.cs [hunk 1]
```diff
@@ -82,6 +82,12 @@ public class APCPoweredDevice : NetworkBehaviour, IServerDespawn, IMultitoolSlav
 		[SyncVar(hook = nameof(UpdateSynchronisedVoltage))]
 		private float recordedVoltage = 0;
 
+		public float Voltage => RelatedAPC == null ? 0 : RelatedAPC.Voltage;
+
+		[SerializeField]
+		private MultitoolConnectionType conType = MultitoolConnectionType.APC;
+		public MultitoolConnectionType ConType => conType;
+
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Engineering/APCPoweredDevice.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs [hunk 2]
```diff
@@ -365,6 +373,36 @@ private IEnumerator DelayMicrowaveRunningSFX()
...
+
+		public void RefreshParts(IDictionary<GameObject, int> partsInFrame)
+		{
+			EnsureInit();
+
+			// Get the machine stock parts used in this instance and get the tier of each part.
+			ICollection<GameObject> parts = partsInFrame.Keys;
+
+			// Collection is unorganized so run through the whole list.
+			foreach (GameObject part in parts)
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs]:
```suggestion
			// Get the machine stock parts used in this instance and get the tier of each part.
			// Collection is unorganized so run through the whole list.
			foreach (GameObject part in partsInFrame.Keys)
```

File: UnityProject/Assets/Scripts/Items/Tool/Multitool/Multitool.cs [hunk 3]
```diff
@@ -92,21 +92,32 @@ public void PrintElectricalThings(PositionalHandApply interaction)
...
+							? $"{device.gameObject.ExpensiveName()}</b> is not connected to an APC!"
+							: $"<b>{device.gameObject.ExpensiveName()}</b>: {device.Wattusage.ToEngineering("W")} " +
+									$"({device.RelatedAPC.Voltage.ToEngineering("V")})");
+				}
+				foreach (var node in electricalNodes)
+				{
+					sb.AppendLine(node.ShowInGameDetails());
+				}
+
+				message = sb.ToString() + "<i>";
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Tool/Multitool/Multitool.cs]: Why not append the closure of the message and only use ToString when displaying the text?

File: UnityProject/Assets/Scripts/Items/Kitchen/Cookable.cs [hunk 4]
```diff
@@ -60,4 +63,12 @@ private void OnBurnUpServer(DestructionInfo info)
 			Spawn.ServerPrefab(CookedProduct, gameObject.RegisterTile().WorldPosition, transform.parent);
 		}
 	}
+
+	[Flags]
+	public enum CookSource
+	{
+		BurnUp = 1,
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/Kitchen/Cookable.cs]: Use the shifting method to assign order, I’m on my phone right now but I recall it lookin like this
```
None = 0,
Something = 1 << 1,
Other = 1 << 2,
Another = 1 << 3,
Final = 1 << 4
```
All flag is created automagically.

**Comment by corp-0** [comment]
I sort of feel like having all states of the microwave declared inside the class/same file, makes it feel more complex than it really is.

**Comment by NoooneyDude** [comment]
> I sort of feel like having all states of the microwave declared inside the class/same file, makes it feel more complex than it really is.

I hear ya. I just didn't see the need to have another public class declared in the namespace, particularly when the file isn't much over my arbitrary ~500 line target and you can just collapse that entire code region. I suppose ideally you'd have these classes declared in another file with scope `internal`, if we used assemblies. Or perhaps create a new namespace just for the microwave - not a bad idea too.

If you insist on a new file though, I'll change it.

**Review by corp-0 — DISMISSED** [review]

File: UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs [hunk 5]
```diff
@@ -479,7 +522,8 @@ public MicrowaveRunning(Microwave microwave)
 				StateMsgForExamine = "running";
 				microwave.OnSyncScreenGlow(microwave.screenGlowEnabled, true);
 				microwave.OnSyncOvenGlow(microwave.ovenGlowEnabled, true);
-				microwave.spriteHandler.ChangeSprite(2);
+				microwave.spriteHandler.ChangeSprite((int) SpriteState.Running);
+				microwave.SetWattage(850);
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs]: maybe change this to a  field you can edit in the editor

**Review by Bod9001 — DISMISSED** [review]

File: UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs (hunk 5 above)
Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs]: Ah, could do. You want one for all states or just the running one?

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Objects/Machines/Microwave.cs]: might as well all states

**Review by Bod9001 — APPROVED** [review]

**Review by corp-0 — APPROVED** [review]

**Pull request merged**